### DISCUSSION TITLE
How to use x264 encoder on AMS on version 2.5.1 and above?

### DISCUSSION THREAD

**Discussion by Mohit-3196** [ant-media/Ant-Media-Server#5087] [FAQs] [answered]
How can I configure AMS to use the x264 encoder on version 2.5.1 and above?

**Comment by yashtandon113** [comment]
AMS uses the **libopenh264** encoder by default starting with v2.5.1. Prior to v2.5.1, the default encoder was **libx264**. Please follow the steps below to utilise the x264 encoder in subsequent versions.

1. Get the FFMPEG jar file.

- Download the file for version 2.5.1 and 2.5.3 from [here](https://drive.google.com/file/d/1D4dxVkvZQOnz1YAhUnGT6KbPvNYs9WBQ/view?usp=share_link).

- Download the jar file for version 2.6.0 and higher from [here](https://drive.google.com/file/d/1g_PwxoqmW1aKootNQVhCMWW7F10GVPVb/view?usp=share_link).

2. Now, copy the jar file to your AMS and place it in the `/usr/local/antmedia/plugins` folder, and grant the antmedia user access to the file.

   `sudo chown -R antmedia:antmedia /usr/local/antmedia/plugins`

3. After copying the file, add the property listed below to the'red5-web.properties' file in the '/usr/local/antmedia/webapps/App-Name/WEB-INF' folder.

   `settings.encoding.encoderName=libx264`

   Please restart the server after making the modifications.

4. After publishing the stream, you should see the following logs.
   
`   2023-05-10 23:47:50,243 [vert.x-worker-thread-24] INFO  i.a.e.adaptive.video.H264Encoder - encoder name: libx264 for stream:test
`

**Reply by muratugureminoglu** [reply]
Don't forget to add ABR (to see libx264 in the log file.) :)

**Reply by mekya** [reply]
ffmpeg gpl builds are also available in the maven repositories. 
https://mvnrepository.com/artifact/org.bytedeco/ffmpeg

**Comment by muratugureminoglu** [comment]
Here is a script to convert libopenh264 to libx264

```
#!/bin/bash
#
# Change libopenh264 to libx264 for all apps
#

v25="https://antmedia.io/libx264/ffmpeg-linux-x86_64-gpl.jar"
v26="https://antmedia.io/libx264/ffmpeg-5.1.2-1.5.8-linux-x86_64-gpl.jar"
PLUGIN_DIR="/usr/local/antmedia/plugins"
AMS_BASE="/usr/local/antmedia/"

VERSION=$(unzip -p $AMS_BASE/ant-media-server.jar | grep -a "Implementation-Version"|cut -d' ' -f2 | tr -d '\r')


if [ "$(printf '%s\n' "2.6" "$VERSION" | sort -V | head -n1)" = "2.6" ]; then
	wget $v26 -P $PLUGIN_DIR
else
	wget $v25 -P $PLUGIN_DIR
fi

chown -R antmedia.antmedia /usr/local/antmedia/plugins/

LIST_APPS=`ls -d $AMS_BASE/webapps/*/`

for i in $LIST_APPS; do
	echo "" >> $i/WEB-INF/red5-web.properties
	echo "settings.encoding.encoderName=libx264" >> $i/WEB-INF/red5-web.properties
done

systemctl restart antmedia
```

**Comment by mekya** [comment, marked as answer]
Let me update the answer with the Ant Media Server v2.14.0


- Download the ffmpeg-7.1-1.5.11-linux-x86_64-gpl.jar [from Maven Repository](https://mvnrepository.com/artifact/org.bytedeco/ffmpeg/7.1-1.5.11) or ask help to get the latest version from support team 

- Copy the jar file to your to `/usr/local/antmedia/plugins` folder, and grant the antmedia user access to the file.
```
sudo chown -R antmedia:antmedia /usr/local/antmedia/plugins
```
- Remove the old file ffmpeg jar
```
sudo rm /usr/local/antmedia/plugins/ffmpeg-7.1-1.5.11-linux-x86_64.jar 
```
- Login the webpanel and go to the advanced settings of application that you use such as `live' and change `encoderName` from empty to `libx264` 
```
 "encoderName": "libx264",
```
- Click the save button and do not forget to add some adaptive bitrates



After publishing the stream, you should see the following logs which means libx264 is working for you
```
[vert.x-worker-thread-24] INFO  i.a.e.adaptive.video.H264Encoder - Video codec opened. .... codecName: libx264 for stream: stream1
```
